npep/npep-126-egress-traffic-control.md

```diff
@@ -0,0 +1,100 @@
+# NPEP-126: Add northbound traffic support in (B)ANP API
+
+* Issue: [#126](https://github.com/kubernetes-sigs/network-policy-api/issues/126)
+* Status: Provisional
+
+## TLDR
+
+This NPEP proposes adding support for cluster egress (northbound) traffic control in the `AdminNetworkPolicy` and `BaselineAdminNetworkPolicy` API objects.
+
+## Goals
+
+* Implement egress traffic control towards external destinations (outside the cluster)
+* Implement egress traffic control towards cluster nodes
+  - Currently the behaviour for policies defined around traffic from cluster
+    workloads (non-hostNetworked pods) towards nodes in the
+    cluster is undefined. See https://github.com/kubernetes-sigs/network-policy-api/issues/73.
+    - ANP currently supports only east-west traffic and this traffic flow cuts from
+    overlay to underlay which makes this part of the egress (northbound) use case.
+    - Let's provide a defined behaviour in ANP to explicitly achieve the use case.
+    - NOTE1: Traffic towards nodes here includes traffic towards host-networked pods on that node
+      because a "node" resource encompasses all objects that share the host-networking resources
+    - NOTE2: Traffic towards nodes here exluding traffic towards host-networked pods is outside
+      the scope of this NPEP and is a non-goal
+* Implement egress traffic control towards hostNetworked pods
+  - Currently the behaviour for policies defined around traffic from cluster
+  workloads (non-hostNetworked pods) towards hostNetworked pods in the
+  cluster is undefined. See https://github.com/kubernetes-sigs/network-policy-api/issues/73.
+  - ANP currently supports only east-west traffic and this traffic flow cuts from
+  overlay to underlay which makes this part of the egress (northbound) use case.
+  - Let's provide a defined behaviour in ANP to explicitly achieve the use case.
+  - NOTE: We can keep this goal experimental in case implementors cannot distinguish between
+    a node and host-networked pod
+
+## Non-Goals
+
+* Implementing southbound traffic use cases is outside the scope of this NPEP
+* Implementing specific api-server user stories where the definition of an api-server
+matches neither a nodeSelector not a podSelector or any expressive external destinaton
+is outside the scope of thie NPEP and perhaps warrants a new NPEP.
+* Implementing traffic towards nodes exluding traffic towards host-networked pods on that
+node is outside the scope of this NPEP and is a non-goal
```

Comment:
This is confusingly worded

Reply:
hmm.. I guess what I meant was we will cover cni-pods -> nodes and cni-pods -> host-networked-pods is what we cover.. if ppl ask for node minus host-networked pods i.e traffic should match the node but not the host-networked pod that is not explicitly supported, BUT now that I think of this.. it can be expressed as:

1. higher precedence rule denying traffic towards host-networked pods
2. lower precedence rule accepting traffic towards node

so that should cover this case, I will just remove this point.